### ISSUE TITLE
How to debug JavaScript when using Jering.Javascript.NodeJS (Fetch-Post)

### ISSUE THREAD

**Issue by zydjohnHotmail** [JeringTech/Javascript.NodeJS#114]
Hello:
I recently need to run some Node.JS code to fetch(POST) some web server response in string text.
The following is my situation:
My OS: Windows 10 (latest version)
Node.js: 16.4.0 (x64)
I have installed the following npm packages:
```
C:\Temp>npm list
Node@1.0.0 C:\Temp
`-- node-fetch@2.6.1
```
The following is the content for fetch.js:
`C:\Temp>type fetch2.js`
```js
const fetch = require('node-fetch');
async function fetch_data1()
{
  const url = 'https://sport.betregal.com/pagemethods_ros.aspx/UpdateEvents'
  const body1 = 'requestString=55677001';
  const cookies_now = 'XXXX'
  const res =
await fetch(url, {method:'POST', body: body1, 
headers: { "Accept": "*.*", "sec-fetch-mode": "no-cors", "Access-Control-Allow-Origin": "*", "requesttarget": "AJAXService", "cookie": cookies_now,  "Content-Type": "application/x-www-form-urlencoded" }});
  const data = await res.text();
  console.log(data)
}

fetch_data1();
```
C:\Temp>
I can run my Node.js script (index.js) without any issue:
`C:\Temp> node fetch.js`

I created one C# Console App.
I added the nuget page:
`Install-Package Jering.Javascript.NodeJS -Version 6.0.1`
I added this JS file to the project: (fetch.js)
```js
const fetch = require('node-fetch');

module.exports = (callback, payload) => {
    (async () => {
        const url = 'https://sport.betregal.com/pagemethods_ros.aspx/UpdateEvents'
        const cookies_now = 'XXXX'
        try {
            const res =
                await fetch(url, {
                    method: 'POST', body: payload, headers: {
                        "Accept": "*.*", "sec-fetch-mode": "no-cors", "Access-Control-Allow-Origin": "*",
                        "requesttarget": "AJAXService", "cookie": cookies_now,
                        "Content-Type": "application/x-www-form-urlencoded"
                    }
                });
            const data = await res.text();
            debugger;
            callback(null, data);
        } catch (err) {
            console.error(err);
            callback(err);
        } finally {
            console.log('Done');
        }
    });
}
```
And in my C# main program, I have the following code: (Target .NET 5.0)
```cs
using Jering.Javascript.NodeJS;
using System;
using System.Threading.Tasks;

namespace BetRegalEventsMarketsNode
{
    class Program
    {
        public const string Path_To_JS_File = @"C:\Temp\fetch.js";
        static async Task Main()
        {
            string text_data =
                await StaticNodeJSService.InvokeFromFileAsync<string>(Path_To_JS_File, args: new[] { "requestString=55677001" });
        }
    }
}
```
I can run compile my program without issue.
But when I run my program, I can't see any return data. I tried to debug my program, as I put 'debugger' in fetch.js file, but it seems program not running at all.
Let me know how I can debug my program or if I made any syntax errors in my JavaScript code.
I just use one simple Node Fetch function to post some data and get return some JSON format data back from the web server.
By the way, as the cookies will be expired in about 2 hours, can I pass cookies as a parameter from C# main program every 2 hours?

**Comment by JeremyTCD** [comment]
Hi, have you considered executing your HTTP request using .Net's [HttpClient](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-5.0)?

If you need to use Node, you can find instructions for debugging [here](https://github.com/JeringTech/Javascript.NodeJS#debugging-javascript).
You'll need to convert `Main()` to:

```cs
static async Task Main()
{
    StaticNodeJSService.Configure<OutOfProcessNodeJSServiceOptions>(options => options.TimeoutMS = -1);
    StaticNodeJSService.Configure<NodeJSProcessOptions>(options => options.NodeAndV8Options = "--inspect-brk");
    string text_data = await StaticNodeJSService.InvokeFromFileAsync<string>(Path_To_JS_File, args: new[] { "requestString=55677001" });
}
```

Then you'll need to navigate to `chrome://inspect` in Chrome.  

User-javascript is outside the scope of this library. But from a cursory read through of your code, you might not be executing your async method. I would do this:

```js
(async () {
...
})()
```

I'm not familiar with this syntax (no trailing `()`), not sure if the method actually runs.
```js
(async () {
...
});
```

**Comment by zydjohnHotmail** [comment]
Hello:
To make things simpler, I wrote another Fetch(GET) node JavaScript file:
C:Temp>type fetch_1get.js
const fetch = require('node-fetch');
async function get1()
{
  const url = 'https://sport.betregal.com/methods/masterevents.ashx/getResponsiveMasterEvent?eventID=23399531&mode=0&isYourBetEnabled=false'
  const cookies_now = 'XXXXXX'
  const res =
  await fetch(url, {method:'GET', headers: { "Accept": "*.*", "sec-fetch-mode": "no-cors", "Access-Control-Allow-Origin": "*", "requesttarget": "AJAXService", "cookie": cookies_now }});
  const data = await res.text();
  console.log(data)
}

get1();
C:\Temp>
The above code works in node, then I modified my code in C# main program as in your example:

using Jering.Javascript.NodeJS;
using System.Threading.Tasks;

namespace BetRegalEventsMarketsNode
{
    class Program
    {
        public const string Path_Get_JS_File = @"C:\Temp\Fetch_Get.js";

        static async Task Main()
        {
            StaticNodeJSService.Configure<OutOfProcessNodeJSServiceOptions>(options => options.TimeoutMS = -1);
            StaticNodeJSService.Configure<NodeJSProcessOptions>(options => options.NodeAndV8Options = "--inspect-brk");
            string text_data = 
                await StaticNodeJSService.InvokeFromFileAsync<string>(Path_Get_JS_File, 
                args: new[] { "https://sport.betregal.com/methods/masterevents.ashx/getResponsiveMasterEvent?eventID=23399531&mode=0&isYourBetEnabled=false" });
        }
    }
}

The following is the Fetch_Get.js (C:\Temp\Fetch_Get.js)
const fetch = require('node-fetch');

module.exports = (callback, url) => {
    async () => {
        const cookies_now = 'XXXXXX'
        const res =
            await fetch(url, {
                method: 'GET', headers: {
                    "Accept": "*.*", "sec-fetch-mode": "no-cors", "Access-Control-Allow-Origin": "*", "requesttarget": "AJAXService", "cookie": cookies_now }
            });
        const data = await res.text();
        callback(null, data);
    }
}

But when I run my code, I still didn’t have any result coming back.
Let me know if I made any mistake in my code.
Finally, I would like to use HttpClient, but I found the tricky part is: for one request header: 
"requesttarget": "AJAXService".  
The ordinary HttpClient (C#) doesn’t have such header, HttpClient simply quits when it finds this request header.  But using Node Fetch, such header works.
You can try to play around with this header.  In C# HttpClient, if you omit such header, the returned result is totally different from the Node Fetch request with such header.

**Comment by zydjohnHotmail** [comment]
Hello:
I suspect there are some installation issues for the package.
I have removed all Node.JS packages, and re-installed all Node.JS (Version 16.4.0) on Windows 10, and I created a very simple C# Console Project with the following simple code:

using Jering.Javascript.NodeJS;
using System;
using System.Threading.Tasks;

namespace BetRegalEventsMarketsNode
{
    class Program
    {
        public const string Path_To_JS_File = @"C:\Temp\index.js";
        static async Task Main()
        {
            StaticNodeJSService.Configure<OutOfProcessNodeJSServiceOptions>(options => options.TimeoutMS = -1);
            StaticNodeJSService.Configure<NodeJSProcessOptions>(options => options.NodeAndV8Options = "--inspect-brk");
            string text_data = await StaticNodeJSService.InvokeFromFileAsync<string>(Path_To_JS_File);
        }
    }
}

The following is the index.js file:
module.exports = (callback) => {
    async () => {
        const data = 'Hello World';
        callback(null, data);
    }
}

My code is very basic bare bone, can’t be more basic, but when I run my code, I didn’t get any results back in C# main program.
Any ideas on how to find if there is any installation issue or configuration issues?
I installed Node.js (Version 16.4.0), node-fetch(Version 2.6.1), I am using Visual Studio 2019 Version 16.10.2, Jering.Javascript.NodeJS (Nuget version 6.0.1)
I also add Node.js to system PATH variable.
But my code simply not working.
Any suggestions?
Thanks,
